### PR TITLE
[Tutorials] Vision Transformer by vit_pytorch tutorial

### PR DESCRIPTION
**Pull request by ViktoriiaRomanova**
This PR adds example which shows OpenFL ability to work with [vit-pytorch](https://github.com/lucidrains/vit-pytorch/) library (vit-pytorch is the library which implements Vision Transformer architecture).

### CODE REVIEW

**Comment by ViktoriiaRomanova** [comment]
I have read the CLA Document and I hereby sign the CLA

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/director/director_config.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+settings:
+  listen_host: localhost
+  listen_port: 50051
+  sample_shape: ['300', '300', '3']
+  target_shape: ['1']
+  #envoy_health_check_period: 5  # in seconds
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/director/director_config.yaml]: Could you please delete commented line

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/envoy_config_one.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+params:
+  cuda_devices: []
+  
+optional_plugin_components:
+ cuda_device_monitor:
+   template: openfl.plugins.processing_units_monitor.pynvml_monitor.PynvmlCUDADeviceMonitor
+   settings: []
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/envoy_config_one.yaml]: Sould we have `cuda_device_monitor` when we have no `cuda_devices`?

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/start_envoy.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -e
+sh check_dataset.sh
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/start_envoy.sh]: In future we are going to refuse `start_envoy.sh`. Could we put this functionality to `DogsCatsShardDescriptor`?

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+class DogsCatsShardDescriptor(ShardDescriptor):
+    """Shard descriptor class."""
+
+    def __init__(self, data_folder: str = 'data',
+                 rank_worldsize: str = '1,3',
+                 enforce_image_hw: str = None) -> None:
+        """Initialize DogsCatsShardDescriptor."""
+        super().__init__()
+        # Settings for sharding the dataset
+        self.rank, self.worldsize = map(lambda x: int(x), rank_worldsize.split(','))
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]:
```python
map(lambda x: int(x), rank_worldsize.split(','))
```
equals to
```python
map(int, rank_worldsize.split(','))
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+        self.data_folder = Path.cwd() / data_folder
+
+        # Settings for resizing data
+        self.enforce_image_hw = None
+        if enforce_image_hw is not None:
+            self.enforce_image_hw = tuple(map(lambda x: int(x), enforce_image_hw.split(',')))
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]:
```python
tuple(map(lambda x: int(x), enforce_image_hw.split(',')))
``` 
equals to
```python
tuple(map(int, enforce_image_hw.split(',')))
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        self.worldsize = worldsize
+        self.dataset_dir = dataset_dir
+        self.enforce_image_hw = enforce_image_hw
+        self.img_path = self.dataset_dir / data_type
+
+        self.img_names = [
+            img_name
+            for img_name in sorted(os.listdir(self.img_path))
+            if img_name[-3:] == 'jpg'
+        ]
```
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: suggestion:
```python
self.img_names = [
    img.name
    for img in self.img_path.iterdir()
    if img.suffix == '.jpg'
]
```

Comment (ViktoriiaRomanova) [reply openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: Good point, but resulted list should also be sorted, to get equivalent datasets at each envoy
```
self.img_names = [
    img.name
    for img in sorted(self.img_path.iterdir())
    if img.suffix == '.jpg'
]
```
But at this point there is no reason to use `iterdir()`. Never the less `suffix` is useful, thanks)

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+2. Put a relevant envoy_config in each of the n folders (n - number of envoys which you would like to use, in this tutorial there is two of them, but you may use any number of envoys) and copy other files from `envoy` there as well.
+3. Modify each `envoy` accordingly:
+
+    - At `start_envoy.sh` change env_one to env_two (or any unique `envoy` names you like)
+
+    - Put a relevant envoy_config `envoy_config_one.yaml` or `envoy_config_two.yaml` (or any other config file name consistent to the configuration file that is called in `start_envoy.sh`).
+4. Make sure that you installed requirements for each `envoy` in your virtual environment: `pip install -r sd_requirements.txt`
+5. Run the `director`: execute `start_director.sh` in director
+6. Run the `envoys`: execute `start_envoy.sh` in each envoy folder. If kaggle-API setting are correct the download of the dataset will be started. If this is not the first `envoy` launch then the dataset will be redownloaded only if some part of the data are missing.
+7. Run the notebook in a prepared virtual environment. For more information about preparation virtual environment look **Preparation virtual environment**.
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md]: Could you please add link to `Preparation virtual environment` section, like `[**Preparation virtual environment**](#preparation-virtual-environment)` , like [this](https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown)

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+8. Congratulations! You've started your first federated learning with OpenFL.
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md]: May be this is not first OpenFL experience for someone, should we leave this line?

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+### Preparation virtual environment
+
+* Install virtualenvwrapper (if you haven't)
+
+    `pip install virtualenvwrapper`
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md]: Can we use specific syntax for this code line?
Like:

```sh
pip install virtualenvwrapper
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+
+* Create virtual environment
+
+    `mkvirtualenv --python=python3.8 name_virtual_environment`
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/README.md]: Also, can we use  specific syntax in all code lines?

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+        self.img_names = self.img_names[self.rank - 1::self.worldsize]
+        # Shuffling the results dataset after choose half pictures of each class
+        shuffle(self.img_names)
+
+    def __getitem__(self, index):
+        """Return a item by the index."""
+        name = self.img_names[index]
+        # Reading data
+        img = Image.open(self.img_path / name)
+        img_class = 1 if name[:3] == 'dog' else 0
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py:54]: Just a little suggestion
```suggestion
        img_class = 1 if name.startswith('dog') else 0
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+        assert name[:3] in {'cat', 'dog'}, 'Wrong object classification'
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py:55]: Just a little suggestion
```suggestion
        assert name.startswith(('cat', 'dog')), 'Wrong object classification'
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+        """Return the len of the dataset."""
+        return len(self.img_names)
+
+
+class DogsCatsShardDescriptor(ShardDescriptor):
+    """Shard descriptor class."""
+
+    def __init__(self, data_folder: str = 'data',
+                 rank_worldsize: str = '1,3',
+                 enforce_image_hw: str = None) -> None:
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: Wrong typing, `enforce_image_hw` is `str` or `None`? It seems like we have to use `Optional`
 ```Python3
from typing import Optional
enforce_image_hw: Optional[str] = None
```
Also, please check this variable typing in ` DogsCatsShardDataset` init method

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+
+from openfl.interface.interactive_api.shard_descriptor import ShardDataset
+from openfl.interface.interactive_api.shard_descriptor import ShardDescriptor
+
+
+class DogsCatsShardDataset(ShardDataset):
+    """Dogs and cats Shard dataset class."""
+
+    def __init__(self, data_type: str, dataset_dir: Path,
+                 rank=1, worldsize=1, enforce_image_hw=None):
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]:
```suggestion
                 rank: int=1, worldsize: int = 1, enforce_image_hw=None):
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+    def __len__(self):
+        """Return the len of the dataset."""
+        return len(self.img_names)
+
+
+class DogsCatsShardDescriptor(ShardDescriptor):
+    """Shard descriptor class."""
+
+    def __init__(self, data_folder: str = 'data',
+                 rank_worldsize: str = '1,3',
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py:77]: Should we change `rank_worldsize` default value to "1,2", like in configs files? It is not a big deal, but for uniformity =)

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/workspace/PyTorch_DogsCats_ViT.ipynb [hunk 16]
```diff
@@ -0,0 +1,550 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "efficient_transformer = Linformer(\n",
+    "    dim=128,\n",
+    "    seq_len=49+1,  # 7x7 patches + 1 cls-token\n",
```
Comment (alexey-khorkin) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/workspace/PyTorch_DogsCats_ViT.ipynb]:
```suggestion
    "    seq_len=49 + 1,  # 7x7 patches + 1 cls-token\n",
```

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/start_envoy.sh [hunk 17]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+set -e
+
+fx envoy start -n env_one --disable-tls --envoy-config-path envoy_config_one.yaml -dh 'localhost' -dp 50051
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/start_envoy.sh]: Redundant quotes.

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 18]
```diff
@@ -0,0 +1,188 @@
...
+        """Initialize DogsCatsShardDescriptor."""
+        super().__init__()
+        # Settings for sharding the dataset
+        self.rank, self.worldsize = map(int, rank_worldsize.split(','))
+
+        self.data_folder = Path.cwd() / data_folder
+        if not os.path.exists(self.data_folder):
+            os.mkdir(self.data_folder)
+
+        if not DogsCatsShardDescriptor.check_dataset(self.data_folder):
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: `if not self.check_dataset(self.data_folder):` or make `check_dataset` an instance method and call it without `data_folder` argument.

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+            print('Your dataset is absent or damaged. Downloading ... ')
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: Why you don't use logger?

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py (hunk 15 above)
Comment (ViktoriiaRomanova) [reply openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py:77]: This is on purpose to make sure that we get rank_worldsize from the config

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                zipobj.extractall(self.data_folder)
+
+            os.remove(self.data_folder / 'train.zip')
+
+            with ZipFile(self.data_folder / 'test.zip', 'r') as zipobj:
+                zipobj.extractall(self.data_folder)
+
+            os.remove(self.data_folder / 'test.zip')
+
+            DogsCatsShardDescriptor.save_all_md5(self.data_folder)
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: Make `save_all_md5` an instance method and call it without `data_folder` argument: `self.save_all_md5()`

Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: Looks like `download_dataset` function

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+                    for chunk in iter(lambda: f.read(4096), b''):
+                        md5_calc.update(chunk)
+                    md5_dict[rel_file] = md5_calc.hexdigest()
+        return md5_dict
+
+    @staticmethod
+    def save_all_md5(data_folder):
+        """Save dataset hash."""
+        all_md5 = DogsCatsShardDescriptor.calc_all_md5(data_folder)
+        with open(os.path.join(data_folder, 'dataset.md5'), 'w') as f:
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: It's not md5-file, it's json-file containing information about md5 for each file in the dataset. It should have .json extension.

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/start_envoy.sh (hunk 17 above)
Comment (dmitryagapov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/start_envoy.sh]: `'localhost'` -> `localhost`

**Review by dmitryagapov — CHANGES_REQUESTED** [review]

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+            json.dump(all_md5, f)
+
+    @staticmethod
+    def check_dataset(data_folder):
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: Maybe better rename to something starts with `is_`. `is_dataset_complete` for example.

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+                with open(data_folder / rel_file, 'rb') as f:
+                    for chunk in iter(lambda: f.read(4096), b''):
+                        md5_calc.update(chunk)
+                    md5_dict[rel_file] = md5_calc.hexdigest()
+        return md5_dict
+
+    @staticmethod
+    def save_all_md5(data_folder):
+        """Save dataset hash."""
+        all_md5 = DogsCatsShardDescriptor.calc_all_md5(data_folder)
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: self instead DogsCatsShardDescriptor

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+        with open(os.path.join(data_folder, 'dataset.md5'), 'w') as f:
+            json.dump(all_md5, f)
+
+    @staticmethod
+    def check_dataset(data_folder):
+        """Check dataset integrity."""
+        new_md5 = DogsCatsShardDescriptor.calc_all_md5(data_folder)
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/envoy/dogs_cats_shard_descriptor.py]: self instead DogsCatsShardDescriptor

File: openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/workspace/PyTorch_DogsCats_ViT.ipynb [hunk 25]
```diff
@@ -0,0 +1,550 @@
...
+    "    def __len__(self):\n",
+    "        self.filelength = len(self._dataset)\n",
+    "        return self.filelength\n",
+    "\n",
+    "    def __getitem__(self, idx):\n",
+    "        img, label = self._dataset[idx]\n",
+    "        img_transformed = self.transform(img).numpy()\n",
+    "        return img_transformed, label[0]\n",
+    "\n",
+    "# Now you can implement you data loaders using dummy_shard_desc\n",
```
Comment (aleksandr-mokrov) [inline openfl-tutorials/interactive_api/PyTorch_DogsCats_ViT/workspace/PyTorch_DogsCats_ViT.ipynb]: `your data loaders`

**Review by aleksandr-mokrov — CHANGES_REQUESTED** [review]

**Review by dmitryagapov — APPROVED** [review]

**Review by alexey-khorkin — APPROVED** [review]

**Review by aleksandr-mokrov — APPROVED** [review]

**Comment by alexey-gruzdev** [comment]
Jenkins please retry a build

**Pull request merged**